File: frontend/src/api/timetable.ts
import { request } from './client'

export interface TimetableSlot {
  id: string
  tenant_id: string
  academic_year_id: string
  class_id: string
  section_id: string
  day_of_week: number
  day_name: string | null
  period_number: number
  start_time: string
  end_time: string
  subject: string
  staff_id: string | null
  staff_name: string | null
  room: string | null
  created_at: string
}

export interface WeeklyTimetable {
  class_id: string
  section_id: string
  academic_year_id: string
  class_name: string
  section_name: string
  slots: TimetableSlot[]
}

export interface SlotUpsert {
  academic_year_id: string
  class_id: string
  section_id: string
  day_of_week: number
  period_number: number
  start_time: string
  end_time: string
  subject: string
  staff_id?: string
  room?: string
}

export function getClassTimetable(params: {
  academic_year_id: string
  class_id: string
  section_id: string
}): Promise<WeeklyTimetable> {
  const p = new URLSearchParams(params)
  return request<WeeklyTimetable>(`/timetable/class?${p}`)
}

export function upsertSlot(data: SlotUpsert): Promise<TimetableSlot> {
  return request<TimetableSlot>('/timetable/slots', { method: 'PUT', body: JSON.stringify(data) })
}

export async function deleteSlot(params: {
  academic_year_id: string
  class_id: string
  section_id: string
  day_of_week: number
  period_number: number
}): Promise<void> {
  const p = new URLSearchParams({
    academic_year_id: params.academic_year_id,
    class_id: params.class_id,
    section_id: params.section_id,
    day_of_week: String(params.day_of_week),
    period_number: String(params.period_number),
  })
  await request<void>(`/timetable/slots?${p}`, { method: 'DELETE' })
}

File: backend/models/timetable.py
from datetime import datetime, time
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

DAY_NAMES = {1: "Monday", 2: "Tuesday", 3: "Wednesday", 4: "Thursday", 5: "Friday", 6: "Saturday"}


class SlotUpsert(BaseModel):
    academic_year_id: UUID
    class_id: UUID
    section_id: UUID
    day_of_week: int
    period_number: int
    start_time: str   # "08:00"
    end_time: str     # "08:45"
    subject: str
    staff_id: Optional[UUID] = None
    room: Optional[str] = None


class SlotResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    academic_year_id: UUID
    class_id: UUID
    section_id: UUID
    day_of_week: int
    day_name: Optional[str] = None
    period_number: int
    start_time: str
    end_time: str
    subject: str
    staff_id: Optional[UUID]
    staff_name: Optional[str] = None
    room: Optional[str]
    created_at: datetime


class WeeklyTimetable(BaseModel):
    class_id: UUID
    section_id: UUID
    academic_year_id: UUID
    class_name: str
    section_name: str
    slots: list[SlotResponse]
